docs/recipes/Authentication.md

```diff
@@ -0,0 +1,1447 @@
+:::
+
+Since we're allowing the Supabase client to manage session storage, it will automatically persist changes to the session. We've added an event listener to stop refreshing the session when the application is no longer in the foreground and restart it when it returns to the foreground. The Supabase client will then automatically refresh the session as necessary; one less thing that we'll need to handle manually. 😮‍💨
+
+### Signing Up & Signing In
+
+To keep this simple, we're going to use the same form for both. We'll need to create an `onPress` and `onSubmit` handler for the respective actions that are already stubbed out in the `SignInScreen` we created earlier. You may have more information you'd like to capture (e.g. name, phone number, password confirmation, etc.) when a user signs up. In such a case, create a separate "Sign Up" screen that captures the additional data.
+
+#### Creating Authentication Context & Provider
+
+We're going to be using the session across components, at different depths in our component tree, and with navigation. For this access pattern we'll create a context and provider. This way if that information changes, we'll re-render the entire tree. (e.g. If the user signs out, we'll navigate back to the "Sign In" screen automatically.)
+
+:::tip
+Be careful when using contexts as anything that depends on data in the context is going to cause a re-render when the data changes. This can have performance implications if you're re-rendering the entire tree frequently. In the case of authentication, we want to re-render the entire tree when the session is updated as we may need to navigate to the sign in screen if the user's session expires.
+:::
+
+Let's setup our `AuthContext` with our session state, add our `AuthProvider`, and create a `useAuth` hook that will return the value of our context:
+
+```typescript title="/app/services/auth/useAuth.tsx"
+import React, {
+  createContext,
+  PropsWithChildren,
+  useCallback,
+  useContext,
+  useState,
+} from "react"
+import { Session, supabase } from "./supabase"
+import { AuthResponse, AuthTokenResponsePassword } from "@supabase/supabase-js"
+
+type AuthState = {
+  isAuthenticated: boolean
+  token?: Session["access_token"]
+}
+
+type SignInProps = {
+  email: string
+  password: string
+}
+
+type SignUpProps = {
+  email: string
+  password: string
+}
+
+type AuthContextType = {
+  signIn: (props: SignInProps) => Promise<AuthTokenResponsePassword>
+  signUp: (props: SignUpProps) => Promise<AuthResponse>
+} & AuthState
+
+const AuthContext = createContext<AuthContextType>({
+  isAuthenticated: false,
+  token: undefined,
+  signIn: () => new Promise(() => ({})),
+  signUp: () => new Promise(() => ({})),
+})
+
+export function useAuth() {
```

Comment:
Nice work with this KCD pattern, really like not having useContext scattered everywhere 👍

Reply:
What's the KCD pattern? Unless you meant OCD pattern in which I have lots of those. 😜

Reply:
Kent C Dodds likes this pattern https://kentcdodds.com/blog/how-to-use-react-context-effectively